src/Thorn_Calculus-Core_Bisimilarities.thy

```diff
@@ -1350,9 +1350,148 @@ text \<open>
+  using assms
+  by
+    (subst (asm) (2) repeated_receive_proper_def)
+    (rotate_tac, induction \<alpha> "A \<triangleright> x. (\<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" Q rule: synchronous_transition.induct,
+     fastforce intro: repeated_receive_proper_def)
+
 lemma repeated_receive_idempotency [thorn_simps]:
   shows "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x \<sim>\<^sub>s A \<triangleright>\<^sup>\<infinity> x. \<P> x"
-  sorry
+proof (coinduction rule: synchronous.up_to_rule [where \<F> = "[\<sim>\<^sub>s] \<frown> \<M>"])
+  case (forward_simulation \<alpha> S)
+  then show ?case
+  proof cases
+    case (parallel_left_io \<eta> A' n X P')
+    from \<open>A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>IO \<eta> A' n X\<rparr> P'\<close>
+    have "A \<triangleright> \<star>\<^bsup>n\<^esup> X = IO \<eta> A' n X" and "P' = post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)"
+      by (fastforce elim: repeated_receive_evolution)+
+    with \<open>A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>IO \<eta> A' n X\<rparr> P'\<close> and \<open>\<alpha> = IO \<eta> A' n X\<close>
+    have "A \<triangleright>\<^sup>\<infinity> x. \<P> x \<rightarrow>\<^sub>s\<lparr>A \<triangleright> \<star>\<^bsup>n\<^esup> X\<rparr> post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)" and \<open>\<alpha> = A \<triangleright> \<star>\<^bsup>n\<^esup> X\<close>
+      by (simp_all only:)
+    moreover
+    have "
+      post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n
+      \<sim>\<^sub>s
+      post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+    proof -
+      have "
+        post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x)
+        =
+        post_receive n X \<P> \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n"
+        using post_receive_after_parallel
+        by simp
+      then have "
+        post_receive n X (\<lambda>v. \<P> v \<parallel> A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n
+        \<sim>\<^sub>s
+        post_receive n X \<P> \<parallel> ((A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n \<parallel> (A \<triangleright>\<^sup>\<infinity> x. \<P> x) \<guillemotleft> suffix n)"
+        by (simp only: parallel_associativity)
+      then show ?thesis
+        by (simp only: adapted_after_parallel)
+    qed
```

Comment:
I guess the proof might be easier to understand if it was in calculational-reasoning style.

Reply:
Also, the `by simp` in line 1388 can be replaced by the simpler `unfolding post_receive_def .`.

Reply:
Fixed in d57dfe8706d960f00e2c8007f3d7293e2f253460, which extends the improvements to the whole proof of `repeated_receive_idempotency`. Please verify.

Reply:
Oh, this trick of combining `unfolding` with `..`, with the latter applying reflexivity of `=`, is really neat. I didn’t know that `standard`, which is invoked by `..`, can solve equations of the form `x = x`. I always used `simp (only: …)` in such cases. Maybe we should replace all invocations of `simp only:` by proofs using this style.

Reply:
The `…` and `∼⇩s` in the second proof step should not be on lines of their own (this sort of splitting across lines is only done in our existing code where there is something other than `…`). So `also have "… ∼⇩s` should all be on one line.

Reply:
There shouldn’t be any parentheses in `(A ▹⇧∞ x. 𝒫 x) « suffix n`, as `▹⇧∞` binds stronger than `«`. Please check also whether there are other places where this issue arises.

Reply:
The application of `adapted_after_parallel` should be made part of the second proof step (or become a proof step of its own), since all reasoning should be made explicit as part of the calculational reasoning steps (after `finally`, there should be only `.` with at most some unfolding of definitions).

Reply:
Fixed in 2319419a8e1b96765cd8c40791a68a125f8453bc, which extends the improvements to the whole proof of `repeated_receive_idempotency`. Please verify.

Reply:
Looks good to me.